state/src/action_handler/mod.rs

```diff
@@ -16,15 +16,49 @@
 
 mod hit;
 
-use cmerkle::TrieMut;
+use std::convert::From;
+use std::sync::Arc;
+
+use cmerkle::{TrieError, TrieMut};
 use ctypes::invoice::Invoice;
+use rlp::DecoderError;
 
-use crate::{StateResult, TopLevelState};
+use crate::{StateError, StateResult, TopLevelState};
 
 pub trait ActionHandler: Send + Sync {
+    fn handler_id(&self) -> u64;
     fn init(&self, state: &mut TrieMut) -> StateResult<()>;
-    fn is_target(&self, bytes: &[u8]) -> bool;
-    fn execute(&self, bytes: &[u8], state: &mut TopLevelState) -> Option<StateResult<Invoice>>;
+    fn execute(&self, bytes: &[u8], state: &mut TopLevelState) -> ActionHandlerResult;
+}
+
+pub fn find_handler_for_id<'a>(id: u64, handlers: &'a [Arc<ActionHandler>]) -> Option<&'a Arc<ActionHandler>> {
+    handlers.iter().find(|handler| handler.handler_id() == id)
+}
+
+pub type ActionHandlerResult = Result<Invoice, ActionHandlerError>;
+
+#[derive(Debug)]
```

Comment:
Why do you derive Debug?

Reply:
https://doc.rust-lang.org/std/result/enum.Result.html#impl-1
Debug trait is required for .expect().